File: android/app/src/main/java/com/mart/distribution/demo/ui/pdf/InvoicePdfExporter.kt
package com.mart.distribution.demo.ui.pdf

import android.content.Context
import android.graphics.Color
import android.graphics.Paint
import android.graphics.pdf.PdfDocument
import com.mart.distribution.demo.data.api.dto.InvoiceDocumentDto
import com.mart.distribution.demo.ui.util.formatDecimal
import java.io.File
import java.io.FileOutputStream

object InvoicePdfExporter {
    fun write(
        context: Context,
        doc: InvoiceDocumentDto,
    ): File {
        val pdf = PdfDocument()
        val pageInfo = PdfDocument.PageInfo.Builder(595, 842, 1).create()
        val page = pdf.startPage(pageInfo)
        val canvas = page.canvas
        val titlePaint =
            Paint(Paint.ANTI_ALIAS_FLAG).apply {
                textSize = 20f
                color = Color.BLACK
                isFakeBoldText = true
            }
        val bodyPaint =
            Paint(Paint.ANTI_ALIAS_FLAG).apply {
                textSize = 11f
                color = Color.DKGRAY
            }
        var y = 56f
        canvas.drawText("KNSR Mart — Distribution", 48f, y, titlePaint)
        y += 36f
        canvas.drawText("Invoice ${doc.invoiceNumber}", 48f, y, titlePaint.apply { textSize = 14f })
        y += 28f
        doc.generatedAt?.let {
            canvas.drawText("Issued: $it", 48f, y, bodyPaint)
            y += 20f
        }
        val order = doc.order
        if (order != null) {
            y += 12f
            canvas.drawText("Subtotal (pre-discount): ${formatDecimal(order.totalAmount)}", 48f, y, bodyPaint)
            y += 18f
            canvas.drawText("Discount: ${formatDecimal(order.discountAmount)}", 48f, y, bodyPaint)
            y += 18f
            canvas.drawText("GST: ${formatDecimal(order.gstAmount)}", 48f, y, bodyPaint)
            y += 18f
            bodyPaint.isFakeBoldText = true
            canvas.drawText("Payable: ${formatDecimal(order.finalAmount)}", 48f, y, bodyPaint)
            bodyPaint.isFakeBoldText = false
            y += 28f
            canvas.drawText("Line items", 48f, y, titlePaint.apply { textSize = 13f })
            y += 22f
            order.items?.forEach { line ->
                val name = line.product?.name ?: line.productId
                val row = "$name  ×${line.quantity}  ${formatDecimal(line.finalAmount)}"
                canvas.drawText(row, 52f, y, bodyPaint)
                y += 16f
                if (y > 780f) return@forEach
            }
        }
        pdf.finishPage(page)
        val safeName = doc.invoiceNumber.replace(Regex("[^A-Za-z0-9-]"), "_")
        val out = File(context.cacheDir, "invoice-$safeName.pdf")
        FileOutputStream(out).use { pdf.writeTo(it) }
        pdf.close()
        return out
    }
}

File: ios/Flashmart/Data/InvoicePdfExporter.swift
import UIKit

enum InvoicePdfExporter {
    static func write(invoice: InvoiceDocument) throws -> URL {
        let pageRect = CGRect(x: 0, y: 0, width: 595, height: 842)
        let renderer = UIGraphicsPDFRenderer(bounds: pageRect)
        let data = renderer.pdfData { context in
            context.beginPage()
            let titleAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.boldSystemFont(ofSize: 20),
                .foregroundColor: UIColor.black,
            ]
            let headingAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.boldSystemFont(ofSize: 14),
                .foregroundColor: UIColor.black,
            ]
            let bodyAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 11),
                .foregroundColor: UIColor.darkGray,
            ]

            var y: CGFloat = 56
            "KNSR Mart — Distribution".draw(at: CGPoint(x: 48, y: y), withAttributes: titleAttrs)
            y += 36
            "Invoice \(invoice.invoiceNumber)".draw(at: CGPoint(x: 48, y: y), withAttributes: headingAttrs)
            y += 28
            if let generated = invoice.generatedAt {
                "Issued: \(generated)".draw(at: CGPoint(x: 48, y: y), withAttributes: bodyAttrs)
                y += 20
            }

            if let order = invoice.order {
                y += 12
                "Subtotal (pre-discount): \(FMTheme.inr(order.totalAmount?.doubleValue ?? 0))"
                    .draw(at: CGPoint(x: 48, y: y), withAttributes: bodyAttrs)
                y += 18
                "Discount: \(FMTheme.inr(order.discountAmount?.doubleValue ?? 0))"
                    .draw(at: CGPoint(x: 48, y: y), withAttributes: bodyAttrs)
                y += 18
                "GST: \(FMTheme.inr(order.gstAmount?.doubleValue ?? 0))"
                    .draw(at: CGPoint(x: 48, y: y), withAttributes: bodyAttrs)
                y += 18
                let boldBody: [NSAttributedString.Key: Any] = [
                    .font: UIFont.boldSystemFont(ofSize: 11),
                    .foregroundColor: UIColor.darkGray,
                ]
                "Payable: \(FMTheme.inr(order.finalAmount?.doubleValue ?? 0))"
                    .draw(at: CGPoint(x: 48, y: y), withAttributes: boldBody)
                y += 28
                "Line items".draw(at: CGPoint(x: 48, y: y), withAttributes: headingAttrs)
                y += 22
                for line in order.items ?? [] {
                    let name = line.product?.name ?? line.productId
                    let row = "\(name)  ×\(line.quantity)  \(FMTheme.inr(line.finalAmount?.doubleValue ?? 0))"
                    row.draw(at: CGPoint(x: 52, y: y), withAttributes: bodyAttrs)
                    y += 16
                    if y > 780 { break }
                }
            }
        }

        let safeName = invoice.invoiceNumber.replacingOccurrences(
            of: #"[^A-Za-z0-9-]"#,
            with: "_",
            options: .regularExpression
        )
        let url = FileManager.default.temporaryDirectory.appendingPathComponent("invoice-\(safeName).pdf")
        try data.write(to: url)
        return url
    }
}
